privacyidea/api/lib/prepolicy.py

```diff
@@ -1201,6 +1201,24 @@ def check_external(request=None, action="init"):
     return True
 
 
+def token_serial_from_path_g(request=None, action=None):
+    """
+    This is decorator function fetches the token serial from the route
+    and adds it to the flask g object. It decorates all api functions
+    where serial can be part of the url.
+    """
+    if g.serial is not None:
+        if request.path:
+            # request.path is something like '/token/setpin/SERIAL' or '/token/SERIAL'
+            path_list = [x for x in request.path.split('/') if x]
+            if path_list[0] == "token" and len(path_list) == 2:
```

Comment:
I am wondering, if this logic to identify *where* the potential serial can be located in the path will always work as expected or if we could run into trouble.

Please consider adding the position of the serial as parameter to the decorator like this:

~~~~python
def token_serial_from_path_g(request=None, action=None, serial_position=2)
~~~~

Then we do not have to have any knowledge about the structure of the endpoints within the decorator.

We could use the decorator like this:

~~~~python
@token_blueprint.route('/disable', methods=['POST'])
@token_blueprint.route('/disable/<serial>', methods=['POST'])
@prepolicy(token_serial_from_path_g, request, serial_position=2)
def disable_api(serial=None):
~~~~

This way we would put the logical decision right to the place where the endpoint is decorated and keep the logic out of the ``token_serial_from_path_g``.
This would especially help, if one day we would have other endpoints, were e.g. the serial is located at position 4 or 5. (I am not planning on this, but you never know).

Reply:
that's a very good idea indeed.

Reply:
Ah, we cannot easily pass arbitrary arguments to policy decorators. Lets talk about this.

Reply:
We do this in the event decorator
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/api/token.py#L150

Reply:
The commit https://github.com/privacyidea/privacyidea/pull/2535/commits/e0d18e04fafd70a7510f00d94ebf2fbe11c736ef implements the APIDecorator class which can be used as a pre or a post decorator and accepts options to pass to the decorating function.

https://github.com/privacyidea/privacyidea/pull/2535/commits/e0d18e04fafd70a7510f00d94ebf2fbe11c736ef#diff-2f390b1f7e67d57678da243f6a900cb3ee0fe1a7137732217e0c7d5d87ab9db2